crates/holochain_p2p/src/spawn/actor.rs

```diff
@@ -1121,17 +1106,72 @@ impl actor::HcP2p for HolochainP2pActor {
 
     fn publish(
         &self,
-        _dna_hash: DnaHash,
-        _request_validation_receipt: bool,
-        _countersigning_session: bool,
-        _basis_hash: holo_hash::OpBasis,
+        dna_hash: DnaHash,
+        request_validation_receipt: bool,
+        countersigning_session: bool,
+        basis_hash: holo_hash::OpBasis,
         _source: AgentPubKey,
-        _op_hash_list: Vec<DhtOpHash>,
+        op_hash_list: Vec<DhtOpHash>,
         _timeout_ms: Option<u64>,
-        _reflect_ops: Option<Vec<DhtOp>>,
+        reflect_ops: Option<Vec<DhtOp>>,
     ) -> BoxFut<'_, HolochainP2pResult<()>> {
         Box::pin(async move {
-            tracing::error!("publish is currently a STUB in holochain_p2p--deferring since we'll eventually get stuff on gossip anyways...");
+            use crate::types::event::HcP2pHandler;
+
+            // This single function is/was a weird mix of a bunch of strange
+            // operations.
+            //
+            // - inexplicably `reflect_ops` back to holochain
+            // - actually publish `op_hash_list`
+            // - send a notification to peers that we'd like validation receipts
+            //   - moving this to a separate call
+            // - i have no idea at all what the countersigning bool is for
+
+            // -- handle the bizzarre reflection thing -- //
+
+            if let Some(reflect_ops) = reflect_ops {
+                self.evt_sender
+                    .get()
+                    .ok_or_else(|| HolochainP2pError::other(EVT_REG_ERR))?
+                    .handle_publish(
+                        dna_hash.clone(),
+                        request_validation_receipt,
+                        countersigning_session,
+                        reflect_ops,
+                    )
+                    .await?;
+            }
+
+            let space = dna_hash.to_k2_space();
+
+            let space = self.kitsune.space(space).await?;
+
+            // -- actually publish the op hashes -- //
+
+            let op_hash_list: Vec<OpId> = op_hash_list.into_iter().map(|h| h.to_k2_op()).collect();
+
+            let urls: std::collections::HashSet<Url> = space
+                .peer_store()
+                .get_near_location(basis_hash.get_loc(), usize::MAX)
+                .await?
+                .into_iter()
+                .filter_map(|info| {
+                    if info.is_tombstone {
+                        return None;
+                    }
+                    info.url.clone()
+                })
+                .collect();
+
+            for url in urls {
+                space
+                    .publish()
+                    .publish_ops(op_hash_list.clone(), url)
+                    .await?;
+            }
+
+            // -- TODO why is there a bool about countersigning?? -- //
```

Comment:
So that when you're receiving a published countersigned entry, you're not treating it as the other ops, i.e. integrate it, because you're awaiting signatures from all parties?

Reply:
Well, that's going to need a separate call or something, because there is no longer any method for passing this bool around with the publish process.

Reply:
Yeah that's right, when you receive the publish it checks this boolelan, and then either triggers the `incoming_dht_ops_workflow` or `receive_incoming_countersigning_ops` workflow. Seems like we may want to split out `PublishCountersigning` as a separate event.

Reply:
Yes please, there are so many places where we have chunky functions that handle both and it makes following countersigning even harder than it needs to be